aptos-move/aptos-workspace-server/src/main.rs

```diff
@@ -0,0 +1,176 @@
+    node_config.storage.enable_indexer = true;
+
+    node_config
+        .api
+        .address
+        .set_ip(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)));
+    node_config
+        .indexer_grpc
+        .address
+        .set_ip(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)));
+
+    node_config.admin_service.address = "127.0.0.1".to_string();
+    node_config.inspection_service.address = "127.0.0.1".to_string();
+
+    let (api_port_tx, api_port_rx) = oneshot::channel();
+    let (indexer_grpc_port_tx, indexer_grpc_port_rx) = oneshot::channel();
+
+    let run_node = {
+        let test_dir = test_dir.to_owned();
+        let node_config = node_config.clone();
+        move || -> Result<()> {
+            start_and_report_ports(
+                node_config,
+                Some(test_dir.join("validator.log")),
+                false,
+                Some(api_port_tx),
+                Some(indexer_grpc_port_tx),
+            )
+        }
+    };
+
+    let _node_thread_handle = thread::spawn(move || {
+        let res = run_node();
+
+        if let Err(err) = res {
+            println!("Node stopped unexpectedly {:?}", err);
+        }
+    });
+
+    let api_port = api_port_rx.await?;
+    let indexer_grpc_port = indexer_grpc_port_rx.await?;
+
+    let api_health_checker = HealthChecker::NodeApi(
+        Url::parse(&format!(
+            "http://{}:{}",
+            node_config.api.address.ip(),
+            api_port
+        ))
+        .unwrap(),
+    );
+    let indexer_grpc_health_checker = HealthChecker::DataServiceGrpc(
+        Url::parse(&format!(
+            "http://{}:{}",
+            node_config.indexer_grpc.address.ip(),
+            indexer_grpc_port
+        ))
+        .unwrap(),
+    );
+
+    api_health_checker.wait(None).await?;
+    eprintln!(
+        "Node API is ready. Endpoint: http://127.0.0.1:{}/",
+        api_port
+    );
+
+    indexer_grpc_health_checker.wait(None).await?;
+    eprintln!(
+        "Transaction stream is ready. Endpoint: http://127.0.0.1:{}/",
+        indexer_grpc_port
+    );
+
+    let faucet_run_config = RunConfig::build_for_cli(
+        Url::parse(&format!(
+            "http://{}:{}",
+            node_config.api.address.ip(),
+            api_port
+        ))
+        .unwrap(),
+        "127.0.0.1".to_string(),
+        0,
+        FunderKeyEnum::KeyFile(test_dir.join("mint.key")),
+        false,
+        None,
+    );
+
+    let (faucet_port_tx, faucet_port_rx) = oneshot::channel();
+    tokio::spawn(faucet_run_config.run_and_report_port(faucet_port_tx));
+
+    let faucet_port = faucet_port_rx.await?;
+
+    let faucet_health_checker =
+        HealthChecker::http_checker_from_port(faucet_port, "Faucet".to_string());
+    faucet_health_checker.wait(None).await?;
+    eprintln!(
+        "Faucet is ready. Endpoint: http://127.0.0.1:{}",
+        faucet_port
+    );
+
+    eprintln!("Indexer API is ready. Endpoint: http://127.0.0.1:0/");
+
+    Ok(())
+}
```

Comment:
Overall this seems like a lot of duplication of what we do in localnet, but without all of the functionlity. For example:
- Localnet already has support for running all of these services.
- Localnet has greater configurability, e.g. you can opt in/out of running the faucet, the txn stream, the indexer API, etc.
- Support for handling ctrl+C gracefully.
- Generic support for post-alive steps, e.g. applying Hasura metadata. 
- Shutdown steps.
- Most critically, it has support for indexer API already, which has a lot of complexity related to how we make it possible for different Docker services to work together on different OSes / Docker platforms, applying DB migrations safely, running Hasura migrations, etc etc.

So at a high level, I wonder why we're doing it all again here? Where the alternative would be to just extract localnet into its own crate and use that. The main difference in functionality I see is the ability to run on unassigned ports and find their ports later, but couldn't that just be added to localnet? Same thing with using a tempdir for the `.testnet` directory.

Reply:
@banool remember we discussed this already on slack?
- You are totally correct that Localnet already has all the components wired up.
- However, exactly due to the way it handles the post-alive and shutdown steps, it requires a non-trivial refactoring to support unassigned ports
  - Specifically, stuff like post-alive and shutdown steps need to be returned as (part of) a future that can only be resolved after the actual ports have been bound
- Based on our previous discussion, we decided to work on a separate "localnet v2" implementation for the short term. 
  - This decoupling offers two major benefits
    - It avoids unexpected disruptions to the existing CLI users of Localnet while we iterate over the implementation
    - It allows Aptos Workspace to evolve faster
- Eventually when the v2 matures, we will use it to replace the original implementation.

This will likely come in the next PR so stay tuned!